### PR TITLE
[Rewrite] Optimize exp(x) - 1 to expm1(x)

### PR DESCRIPTION
**Pull request by aimanalhazmi**
- Add an exp_minus_one rewrite rule to fuse exp(x) - 1 into expm1(x).
- Enable the new rule by default.
- Add strict guards to prevent matching incorrect constants or reversed operations.
- Add tests for edge cases, disabled configuration, and rewrite-ordering sensitivity.

### CODE REVIEW

**Comment by aimanalhazmi** [comment]
This PR is based on https://github.com/deem-data/stratum/pull/120, so its diff includes both rewrites. You can either merge this as-is with both rewrites, or merge https://github.com/deem-data/stratum/pull/120 first. I'll then rebase this PR onto main, and you can merge it afterward.

**Comment by aimanalhazmi** [comment]
> This PR is based on #120, so its diff includes both rewrites. You can either merge this as-is with both rewrites, or merge #120 first. I'll then rebase this PR onto main, and you can merge it afterward.

Done

File: stratum/optimizer/_algebraic_rewrites.py
Comment (e-strauss) [inline stratum/optimizer/_algebraic_rewrites.py:47]: I would add this rewrite here at this position (the order of the rewrites is important):
so that we make the fused `expm1` operator before we do the other rewrite checks.
currently you do the fusing after we did the other rewrites, which causes for example:
```python
# out = log1p(exp(df) - 1)
df = st.as_data_op(0)
t1 = df.skb.apply_func(np.exp)
t2 = t1 - 1
t3 =  t3.skb.apply_func(np.log1p)
out, *_ = optimize(t3)
# should only contain one operator since we first rewrite :
# out = log1p(expm1(df)) # existing eliminate_expm1_log1p(root) matches
# out = df
```
I would not put rewrite before the normal exp / log rewrites, since it might block this rewrite:
exp(log(x)) - 1 which should be rewritten to x - 1, but if we ran first the expm1 rewrite we get
expm1(log(x)) and don't match anymore. 

If we also could this two cases in the tests would be cool

File: stratum/optimizer/_numeric_rewrites.py [hunk 1]
```diff
@@ -78,6 +78,16 @@ def action(op1: Op, op2: Op, root: Op) -> Op:
...
 
+def match_exp_minus_one(op):
+    """Match exp(x) - 1"""
+    if isinstance(op, NumericOp) and op.type is NumericOpType.EXP and len(op.outputs) == 1:
+        op2 = op.outputs[0]
+        if (isinstance(op2, NumericOp) and op2.type is NumericOpType.SUBTRACT
+                and op2.opt_operand is None and op2.constant == 1 and not op2.reversed):
+            return (op, op2)
+    return None
+
```
Comment (e-strauss) [inline stratum/optimizer/_numeric_rewrites.py:89]: adding a new rule is fine, since we can't reuse any existing helper in this case

File: stratum/tests/logical_optimizer/algebraic_rewrites/test_numeric.py
Comment (e-strauss) [inline stratum/tests/logical_optimizer/algebraic_rewrites/test_numeric.py:1]: tests look good to me, please see the earlier comment

**Review by e-strauss — COMMENTED** [review]
@aimanalhazmi  thanks for working on this, please review my comments

File: stratum/optimizer/_algebraic_rewrites.py
Comment (aimanalhazmi) [reply stratum/optimizer/_algebraic_rewrites.py:47]: Thank you for the review. That's right, I did not think about it. Thanks!

**Comment by aimanalhazmi** [comment]
Can you please review it again @e-strauss

**Comment by e-strauss** [comment]
> Can you please review it again @e-strauss

LGTM, thanks

**Review by e-strauss — APPROVED** [review]
